0038-http-auth-profiles/http-auth-profiles.md

```diff
@@ -0,0 +1,191 @@
+
+## Specification
+
+### Overview
+This document defines three authentication profiles for securing HTTP network connections in Interledger.
+
+### Authentication
+When authenticating requests between Interledger nodes, it is important to choose an algorithm that maintains an appropriate balance between usability and security, while at the same time maintaining high-performance _and_ interoperability.
+
+In order to find this balance, this document defines three Authentication profiles, each with various trade-offs that should be considered before use:
+
+* `SIMPLE`: Allows two ILP nodes to utilize a previously agreed-upon shared-secret as a [Bearer token](https://tools.ietf.org/html/rfc6750) in all HTTP requests. Peers SHOULD consider this token to be opaque and SHOULD NOT derive any special meaning from the token. 
+
+* `JWT_HS_256`: Allows two ILP nodes to utilize a previously agreed-upon shared-secret in order to _derive_ a `Bearer token` that conforms to the JSON Web Token (JWT) specification as defined in [RFC-7519](https://www.rfc-editor.org/rfc/rfc7519.html) using the `HS_256` signing algorithm defined in section 3.2 of [RFC-7518](https://www.rfc-editor.org/rfc/rfc7518#section-3.2).
+
+* `JWT_RS_256`: Allows two ILP nodes to utilize public-key pairs to _derive_ a `Bearer token` that conforms to the JSON Web Token (JWT) specification as defined in [RFC-7519](https://www.rfc-editor.org/rfc/rfc7519.html) using the `RS_256` signing algorithm defined in section 3.3 of [RFC-7518](https://www.rfc-editor.org/rfc/rfc7518#section-3.3).
+
+Peers MAY use any standard HTTP authentication mechanism to authenticate incoming requests, but SHOULD support `JWT_HS_256` at a minimum. It is RECOMMENDED to use `JWT_RS_256` for production deployments.
+
+#### `SIMPLE` Authentication Profile
+This profile allows two ILP nodes to utilize a previously agreed-upon shared-secret that contains at least 32 bytes (256 bits) of randomly generated data, and is encoded using Base64.
+
+Because tokens in this profile do not inherently contain information about the identity of the caller, requests MUST contain an additional HTTP request-header named `Auth-Principal`.
+ 
+This extra header allows for the identity of the authentication request to be separated from authentication token itself, which reduces computational overhead as well as data-management complexity (e.g., implementations do not need to create data-store indexes using derivations of tokens for lookup purposes).
+
+##### Example Usage
+An example shared-secret in this profile is `HEiMCp0FoAC903QHueY89gAWJHo/izaBnJU8/58rlSI=`. This shared secret is passed as an `Authorization` header in each HTTP request, using the Bearer token scheme, along with an `Auth-Principal` header like this: 
+
+```
+Auth-Principal: alice-usd-123
+Authorization: Bearer HEiMCp0FoAC903QHueY89gAWJHo/izaBnJU8/58rlSI=
+``` 
+
+Implementations MAY support this profile, but SHOULD consider it for development purposes only.
```

Comment:
Why? Plenty of services use this with real money today

Also, this doesn't specify how that token is created, so it could be the result of an OAuth flow. The defining characteristic of this profile is that the token is opaque to one party

Reply:
> Why?

See trade-offs and comments elsewhere. It's a trade-off, but I consider non-expiring tokens to be a very difficult security surface to defend against.

Reply:
> Also, this doesn't specify how that token is created, so it could be the result of an OAuth flow. The defining characteristic of this profile is that the token is opaque to one party

I'm not clear on what you're suggesting. We should probably first clarify if the token is opaque or not per the other threads.

Reply:
> Loads of services use API keys (including payment companies like Stripe and PayPal) and don't seem to be too worried about the issues you describe with the SIMPLE profile. 

I've adjusted the language in the spec in this regard -- in reality if two peers want to use the SIMPLE profile, they should be able to. In fairness to the `SIMPLE` scheme, I have been biasing towards the use-case of two nodes exchanging large values of payments (e.g., two FI's) and I think this has been a mistake in my assumption model. 

I'm going to resolve this issue since I've removed most of the problematic wording around `SIMPLE` - let's work on defining the details of the `SIMPLE` scheme a bit more clearly though.